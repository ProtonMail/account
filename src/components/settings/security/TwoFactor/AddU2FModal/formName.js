import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';

import styles from './index.css';
import settingsActions from '../../../../../actions/settings';
import { connect } from 'unistore/full/preact';

/**
 * Label form for the U2F key.
 */
const FormName = ({ onSubmit, onCancel, addU2FKeyNameAction, settings: { addU2FKey: addU2FKeyStore } }) => {
    const model = { name: (addU2FKeyStore.response ? addU2FKeyStore.response.name : '') };
    return (<ModalWrapper
        onSubmit={(e) => {
            e.preventDefault();
            addU2FKeyNameAction(model.name);
            onSubmit();
        }}
        onReset={(e) => {
            e.preventDefault();
            onCancel();
        }}>
        <ModalContent>
            <div class={styles.name}>
                <label class={styles.nameLabel} htmlFor="name">
                    Name
                </label>
                <div class={styles.nameInputContainer}>
                    <input
                        onInput={({ target: { value } }) => {
                            model.name = value;
                        }}
                        value={model.name}
                        type="name"
                        id="inputName"
                        required
                        placeholder="Name"
                        class={styles.nameTextInput}
                        ref={input => input && input.focus()}
                    />
                </div>
            </div>
        </ModalContent>
        <ModalFooter>
            <button type="reset" value="Reset">
                Back
            </button>
            <button type="submit" value="Submit">
                Next
            </button>
        </ModalFooter>
    </ModalWrapper>);
};

export default connect('settings', settingsActions)(FormName);
