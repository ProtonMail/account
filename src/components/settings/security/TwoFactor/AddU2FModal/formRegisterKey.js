import { Component } from 'preact';
import { connect } from 'unistore/full/preact';
import settingsActions from '../../../../../actions/settings';

import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';
import TextButton from '../../../../ui/TextButton';

import styles from './index.css';

import image from './sign-u2f.png';
import { getErrorMessage } from '../../../../../helpers/u2f';

/**
 * Modal Form to register a new U2F Key.
 *
 * Fetches the challenge from the server, forward it to U2F API, and sends back the answer to the API.
 */
export class FormRegisterKey extends Component {

    /**
     * when next button is pressed.
     * @param {Event} e
     */
    onSubmit(e) {
        e.preventDefault();
        this.props.onSubmit();
    }

    componentDidMount() {
        this.props.addU2FKeyRegisterAction();
    }

    componentWillReceiveProps(newProps) {
        const {
            settings: { addU2FKey: { status: newStatus } }
        } = newProps;

        const {
            settings: { addU2FKey: { status } }
        } = this.props;

        if (newStatus !== status && newStatus === 'finished') {
            this.props.forbidClosure();
        }
    }

    /**
     * Renders the status field.
     * @return {*}
     */
    renderStatus() {
        const {
            settings: { addU2FKey: { response: { name } = {}, status, request, error } }
        } = this.props;

        if (status !== 'failure') {
            return (<div class={styles.status}>
                <div class={styles.row}>
                    <span class={styles.text}>Activate your key</span>
                    <span class={styles.text}>
                            {status || 'fetching'}...
                        </span>
                </div>

                <div class={styles.row}>
                    <span class={styles.text}>Name</span>
                    <span class={[styles.text, styles.nameLabel].join(' ')}>{name}</span>
                </div>
            </div>);

        }
        const { metaData: { code } = {} } = error;
        if (code) {
            return (<div class={styles.status}>
                <span>{getErrorMessage(code, true)}</span>
                <TextButton onClick={() => this.props.addU2FKeyRegisterAction()}>Retry</TextButton>
            </div>);
        }

        this.props.onReset(error.message);
    }

    render () {
        const {
            settings: { addU2FKey: { status } }
        } = this.props;

        return (
            <ModalWrapper
                onSubmit={(e) => this.onSubmit(e)}
                onReset={(e) => {
                    e.preventDefault();
                    this.props.onCancel();
                }}
            >
                <ModalContent class={styles.container}>
                    <img src={image}/>

                    {this.renderStatus()}
                </ModalContent>
                <ModalFooter>
                    <button type="reset" value="Reset" disabled={(status === 'finished')}>
                        Back
                    </button>
                    <button type="submit" value="Submit" disabled={(status !== 'finished')}>
                        Next
                    </button>
                </ModalFooter>
            </ModalWrapper>
        );
    }
}


export default connect(['scope', 'settings'], settingsActions)(FormRegisterKey);
