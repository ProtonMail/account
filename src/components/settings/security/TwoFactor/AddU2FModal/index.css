.name {
    display: flex;
    justify-content: space-between;
    align-self: stretch;
}

.name .label {
    flex: 1;
}

.name .inputContainer {
    flex: 4;
}

.name .textInput {
    width: 100%
}

.RegisterKeyForm-container {
    align-items: stretch;
}

.RegisterKeyForm-status {
    margin: auto 20%;
}

.RegisterKeyForm-status > div {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.RegisterKeyForm-status > div > span {
    flex: 1;
}

.RegisterKeyForm-status > div > span:first-child {
    text-align: left;
}

span.label {
    color: #9592d1;
}
