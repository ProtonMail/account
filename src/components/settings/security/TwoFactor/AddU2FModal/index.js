import Presentation from './presentation';
import FormName from './formName';
import RegisterKeyForm from './formRegisterKey';
import { steps as RecoveryCodeSteps, beforeDismiss as beforeDismissRecoveryCodeModal } from '../SaveRecoveryCodeModal';
import { steps as ScopeModal, beforeDismiss as beforeDismissScopeModal } from '../../../../auth/ScopeModal/index';
import store from '../../../../../helpers/store';
import actions from '../../../../../actions/settings';

export const steps = [
    {
        title: 'Register new U2F Key',
        component: ({ onNextStep, onPreviousStep, message }) => (
            <Presentation onSubmit={onNextStep} onCancel={onPreviousStep} message={message} />
        )
    },
    ...ScopeModal('password'),
    {
        title: 'Name your U2F Key',
        component: ({ onNextStep, onPreviousStep }) => <FormName onSubmit={onNextStep} onCancel={onPreviousStep} />
    },
    {
        title: 'Register new U2F Key',
        component: ({ onNextStep, onPreviousStep, forbidClosure, onReset }) => (
            <RegisterKeyForm
                onSubmit={onNextStep}
                onCancel={onPreviousStep}
                forbidClosure={forbidClosure}
                onReset={onReset}
            />
        )
    },
    ...RecoveryCodeSteps
];

export const beforeDismiss = () => {
    beforeDismissScopeModal();
    beforeDismissRecoveryCodeModal();
    const { resetStoreAction } = actions(store);
    resetStoreAction(store.getState(), ['addU2FKey']);
};
