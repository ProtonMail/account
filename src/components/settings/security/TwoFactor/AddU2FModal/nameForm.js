import { h, Component } from 'preact';
import { styles as ModalStyles } from '../../../../ui/Modal';
import styles from './index.css';

/**
 * Label form for the U2F key.
 */
export default class NameForm extends Component {

    /**
     * triggered each time the name is updated.
     * @param event
     */
    onNameUpdated ( event ) {
        this.setState({ name: event.target.value });
    }

    /**
     * Submit the form.
     * @param e
     */
    onSubmit ( e ) {
        e.preventDefault();
        this.props.onSubmit({ name: this.state.name });
    }

    /**
     * @constructor
     */
    constructor ( props ) {
        super(props);
        this.state = {
            name: ''
        };
    }

    render () {
        return (
            <form
                className={ModalStyles.wrapper}
                onSubmit={this.onSubmit.bind(this)}
                onReset={( e ) => {
                    e.preventDefault();
                    this.props.onCancel({});
                }}
            >
                <div className={ModalStyles.content}>
                    <div class={styles[ 'NameForm-name' ]}>
                        <label htmlFor="name">
                            Name
                        </label>
                        <div>
                            <input
                                onInput={this.onNameUpdated.bind(this)}
                                value={this.state.name}
                                type="name"
                                id="inputName"
                                placeholder="Name"
                            />
                        </div>
                    </div>
                </div>
                <div class={ModalStyles.footer}>
                    <button type="reset" value="Reset">
                        Back
                    </button>
                    <button type="submit" disabled={!this.state.name} value="Submit">
                        Next
                    </button>
                </div>
            </form>
        );
    }
}
