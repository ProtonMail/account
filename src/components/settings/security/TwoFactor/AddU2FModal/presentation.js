import { h } from 'preact';

import { Content as ModalContent, Footer as ModalFooter, Wrapper as ModalWrapper } from '../../../../ui/Modal';
import Link from '../../../../ui/Link';

/**
 * Shows information about the U2F Key usage.
 * @param {Object} props
 * @param {Function} props.onSubmit - triggers the next step.
 * @param {Function} props.onCancel - triggers the previous step.
 * @return {*}
 */
const Presentation = ({ onSubmit, onCancel, message }) => (
    <ModalWrapper
        onSubmit={(e) => {
            e.preventDefault();
            onSubmit();
        }}
        onReset={(e) => {
            e.preventDefault();
            onCancel();
        }}
    >
        {message ? (
            <ModalContent>
                <div>{message}</div>
            </ModalContent>
        ) : (
            <ModalContent>
                <p>This wizard will add a new security key to your Proton Account. </p>
                <p>
                    Please note that you will not be able to access your account if you loose your U2F device and your
                    recovery codes. We recommend setting up a second 2FA method as a backup.
                </p>
                <p>
                    <strong>
                        If you have never used 2FA before, we strongly recommend you reading our 2FA Guide first.
                    </strong>
                </p>
                <div>
                    <Link href="https://protonmail.com/blog/">READ U2F GUIDE </Link>
                </div>
            </ModalContent>
        )}
        <ModalFooter>
            <button type="reset" value="Reset">
                Back
            </button>
            <button type="submit" value="Submit">
                Next
            </button>
        </ModalFooter>
    </ModalWrapper>
);

export default Presentation;
