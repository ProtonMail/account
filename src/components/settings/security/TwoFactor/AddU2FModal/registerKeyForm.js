import { h, Component } from 'preact';
import { register } from 'u2f';

import { styles as ModalStyles } from '../../../../ui/Modal';
import styles from './index.css';

import image from './sign-u2f.png';
import appProvider from 'frontend-commons/src/appProvider';

export default class RegisterKeyModal extends Component {
    onRegisterSucceed ( result ) {
        this.setState({ result, status: 'success' });
    }

    onRegisterFails ( result ) {
        this.setState({ result, status: 'failure' });
    }

    onSubmit ( e ) {
        e.preventDefault();
        this.props.onSubmit({
            codes: [
                'd1e4822e',
                '2d83d85b',
                '717ebed4',
                'a9ffee38',
                'f34ebdf7',
                'd1321481',
                '2572f6cd',
                '39bd63b3',
                'c8016641',
                'a9b4cf9d',
                '8f475f77',
                '123d33b1'
            ]
        });
    }

    constructor () {
        super();
        this.state = {
            status: 'initialize',
            request: {
                Challenge: 'KGOcAJhmPwMKZ4r8vFb2vZZktLh9wZCmLKHxQQH1bxY',
                Versions: [ 'U2F_V2' ],
                RegisteredKeys: []
            }
        };
    }

    componentWillMount () {
        if (this.state.status === 'initialize') {
            const u2fConfig = appProvider.getConfig('u2f');
            register(this.state.request, u2fConfig.appID, u2fConfig.timeout)
                .then(this.onRegisterSucceed.bind(this))
                .catch(this.onRegisterFails.bind(this));
            this.setState({ status: 'waiting' });
        }
    }

    render () {
        const name = this.props.params.name;
        return (
            <form
                class={ModalStyles.wrapper}
                onSubmit={this.onSubmit.bind(this)}
                onReset={( e ) => {
                    e.preventDefault();
                    this.props.onCancel({});
                }}
            >
                <div class={[ styles[ 'RegisterKeyForm-container' ], ModalStyles.content ].join(' ')}>
                    <img src={image}/>

                    <div class={styles[ 'RegisterKeyForm-status' ]}>
                        <div>
                            <span>Activate your key</span>
                            <span>
                                {this.state.status === 'initialize' ? 'fetching' : this.state.status}...
                            </span>
                        </div>

                        <div>
                            <span>Name</span>
                            <span class={styles.label}>{name}</span>
                        </div>
                    </div>
                </div>
                <div class={ModalStyles.footer}>
                    <button type="reset" value="Reset">
                        Back
                    </button>
                    <button type="submit" value="Submit" disabled={this.state.status !== 'success'}>
                        Next
                    </button>
                </div>
            </form>
        );
    }
}
