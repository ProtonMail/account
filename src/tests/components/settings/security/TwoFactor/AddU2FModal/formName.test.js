import render from 'preact-render-to-string';
import { deep } from 'preact-render-spy';

import { FormName } from '../../../../../../components/settings/security/TwoFactor/AddU2FModal/formName';

describe('AddU2FModal SharedSecret step', () => {
    test('display', () => {
        expect(render(<FormName settings={{ addU2FKey: {} }}/>)).toMatchSnapshot();
        expect(render(<FormName settings={{ addU2FKey: { response: { label: 'Test for key' } } }}/>)).toMatchSnapshot();
    });

    test('input event', () => {
        const context = deep(<FormName settings={{ addU2FKey: {} }}/>);
        context.find('input').simulate('input', { target: { value: 'Something' } });
    });

    test('submit and cancel', () => {
        const onSubmit = jest.fn();
        const onCancel = jest.fn();
        const addU2FKeyLabelAction = jest.fn();
        const context = deep(<FormName settings={{ addU2FKey: { response: { label: 'Test for key' } } }} {...{
            onSubmit,
            onCancel,
            addU2FKeyLabelAction
        }}/>);

        const event = { preventDefault: () => undefined };

        context.find('form').simulate('submit', event);
        expect(onSubmit).toHaveBeenCalled();
        expect(addU2FKeyLabelAction).toHaveBeenLastCalledWith('Test for key');
        expect(onCancel).not.toHaveBeenCalled();

        context.find('form').simulate('reset', event);
        expect(onSubmit).toHaveBeenCalledTimes(1); // the submit event
        expect(addU2FKeyLabelAction).toHaveBeenCalledTimes(1);
        expect(onCancel).toHaveBeenCalled(); // only once
    });
});
