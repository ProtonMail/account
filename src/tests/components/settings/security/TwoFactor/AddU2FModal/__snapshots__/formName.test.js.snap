// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddU2FModal FormName step display 1`] = `
<form class="wrapper">
  <div class="content ">
    <div class="name">
      <label class="nameLabel" htmlFor="name">Name</label>
      <div class="nameInputContainer">
        <input value type="name" id="inputName" required placeholder="Name" class="nameTextInput"
        />
      </div>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </div>
</form>
`;

exports[`AddU2FModal FormName step display 2`] = `
<form class="wrapper">
  <div class="content ">
    <div class="name">
      <label class="nameLabel" htmlFor="name">Name</label>
      <div class="nameInputContainer">
        <input value="Test for key" type="name" id="inputName" required placeholder="Name"
        class="nameTextInput" />
      </div>
    </div>
  </div>
  <div class="footer">
    <button type="reset" value="Reset">Back</button>
    <button type="submit" value="Submit">Next</button>
  </div>
</form>
`;
